desc/integrals/_bounce_utils.py

```diff
@@ -700,127 +845,315 @@ def interp_to_argmin_hard(h, points, knots, g, dg_dz, method="cubic"):
+    g(ϑ(α=α₀,ζ), ζ) will sample the approximation to f(α=α₀ ± ε, ζ) with ε → 0 as
+    g converges to f.
+
+    (Visually, the small discontinuity apparent in g(α, ζ) at cuts of the field
+    line will not be visible in g(ϑ, ζ) because when moving along the field line
+    with g(ϑ, ζ) one is continuously flowing away from the starting field line,
+    (whereas g(α, ζ) has to "decide" at the cut what the next field line is).
+    (If full convergence is difficult to achieve, then in the context of surface
+    averaging bounce integrals, function approximation in (α, ζ) coordinates
+    might be preferable because most of the bounce integrals do not stretch
+    across toroidal transits).)
+
+    Note that if g is an unbounded function, as all coordinates are, then
+    it is impossible to approximate it with a finite number of periodic
+    basis functions, so we are forced to use a Fourier Chebyshev series to
+    interpolate θ anyway.
+
+    We explicitly enforce continuity of our approximation of θ between
+    cuts to short-circuit the convergence of the Fourier series for θ.
+    This works to remove the small discontinuity between cuts of the field line
+    because the first cut is on α=0, which is a knot of the Fourier series, and
+    the Chebyshev points include a knot near endpoints, so θ at the next cut of
+    the field line is known with precision.
+
+    """
+    # peeling off field lines
+    fieldline = get_fieldline(alpha, iota, num_transit, 2 * jnp.pi)
+    # Evaluating set of single variable maps is more efficient than evaluating
+    # multivariable map, so we project θ to a set of Chebyshev series.
+    T = FourierChebyshevSeries(f=theta, domain=(0, 2 * jnp.pi)).compute_cheb(fieldline)
+    T.stitch()
+    assert T.X == num_transit
+    assert T.Y == theta.shape[-1]
+    return T
+
+
+def chebyshev(n0, n1, NFP, T, f, Y):
+    """Chebyshev along field lines.
+
+    Parameters
+    ----------
+    T : PiecewiseChebyshevSeries
+        Set of 1D Chebyshev spectral coefficients of θ along field line.
+        {θ_α : ζ ↦ θ(α, ζ) | α ∈ A} where A = (α₀, α₁, …, αₘ₋₁) is the same
+        field line.
+    f : jnp.ndarray
+        Fourier transform of f(θ, ζ) as returned by ``_fourier``.
+        ``n0=grid.num_theta``, ``n1=grid.num_zeta``, ``NFP=grid.NFP``.
+    Y : int
+        Chebyshev spectral resolution for ``f``. Preferably power of 2.
+
+    Returns
+    -------
+    f : PiecewiseChebyshevSeries
+        Set of 1D Chebyshev spectral coefficients of θ along field line.
+        {f_α : ζ ↦ f(α, ζ) | α ∈ A} where A = (α₀, α₁, …, αₘ₋₁) is the same
+        field line.
+
+    """
+    # When f = |B|, it is expected that Y > T.Y so the code immediately below
+    # is then up-sampling the Chebyshev resolution, which is good since the
+    # spectrum of |B| is wider than θ.
+
+    # θ at Chebyshev points, reshaped to (num rho, num transit * num points)
+    theta = T.evaluate(Y).reshape(*T.cheb.shape[:-2], T.X * Y)
```

Comment:
the shape of T.cheb is what? this is just grabbing one index (corresponding to num rho) or is something more subtle going on?

Reply:
T.cheb has shape (..., T.X, T.Y). where ... might be an axis for num rho

Reply:
From Boyd's book:
> The reason that a multi-dimensional BVP
solver may take weeks or months to develop is not because of the intricacies of Sobolev
spaces, but because of the care and fussing that is needed to correctly use the !*#$%& [ex-
pletive deleted] matrix indices.

Reply:
i am reshaping it so that the last axis stores all the points on a particular flux surface, so that i can evaluate those points with the fourier series that is defined on that surface